plonky/src/groth16.rs

```diff
@@ -0,0 +1,91 @@
+use crate::bellman_ce::groth16::{
+    create_random_proof, generate_random_parameters, prepare_verifying_key, verify_proof,
+    Parameters, PreparedVerifyingKey, Proof, VerifyingKey,
+};
+use crate::bellman_ce::Circuit;
+use crate::errors::{EigenError, Result};
+use crate::franklin_crypto::bellman::pairing::Engine;
+use crate::snark::SNARK;
+use rand::Rng;
+use std::marker::PhantomData;
+
+pub struct Groth16<E: Engine, C: Circuit<E>> {
+    _engine: PhantomData<E>,
+    _circuit: PhantomData<C>,
+}
+
+impl<E: Engine, C: Circuit<E>> SNARK<E::Fr> for Groth16<E, C> {
+    type Circuit = C;
+    type AssignedCircuit = C;
+    type ProvingKey = Parameters<E>;
+    type VerificationKey = VerifyingKey<E>;
+    type PreparedVerificationKey = PreparedVerifyingKey<E>;
+    type Proof = Proof<E>;
+
+    fn circuit_specific_setup<R: Rng>(
+        circuit: Self::Circuit,
+        rng: &mut R,
+    ) -> Result<(Self::ProvingKey, Self::PreparedVerificationKey)> {
+        // use std::time::Instant;
+        // let t = Instant::now();
+        let pk = generate_random_parameters::<E, Self::Circuit, R>(circuit, rng)?;
+        let pvk = prepare_verifying_key(&pk.vk);
+        // let elapsed = t.elapsed().as_secs_f64();
+        // log::debug!("groth16 setup run time: {} secs", elapsed);
+        Ok((pk, pvk))
+    }
+
+    fn prove<R: Rng>(
+        circuit_pk: &Self::ProvingKey,
+        input_and_witness: Self::AssignedCircuit,
+        rng: &mut R,
+    ) -> Result<Self::Proof> {
+        // use std::time::Instant;
+        // let t = Instant::now();
+        let result = create_random_proof::<E, _, _, _>(input_and_witness, circuit_pk, rng)?;
+        // let elapsed = t.elapsed().as_secs_f64();
+        // log::debug!("groth16 generate proof run time: {} secs", elapsed);
+        Ok(result)
+    }
+
+    fn verify_with_processed_vk(
+        circuit_pvk: &Self::PreparedVerificationKey,
+        public_input: &[E::Fr],
+        proof: &Self::Proof,
+    ) -> Result<bool> {
+        // use std::time::Instant;
+        // let t = Instant::now();
+        let result = verify_proof(&circuit_pvk, proof, &public_input)?;
+        // let elapsed = t.elapsed().as_secs_f64();
+        // log::debug!("groth16 verify run time: {} secs", elapsed);
+        Ok(result)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::bellman_ce::bn256::Bn256;
+    use crate::{circom_circuit::CircomCircuit, plonk, reader};
+    const CIRCUIT_FILE: &'static str =
+        concat!(env!("CARGO_MANIFEST_DIR"), "/../test/mycircuit.r1cs");
+    #[test]
+    fn groth16_proof() -> Result<()> {
+        let circuit: CircomCircuit<Bn256> = CircomCircuit {
+            r1cs: reader::load_r1cs(CIRCUIT_FILE),
+            witness: None,
+            wire_mapping: None,
+            aux_offset: plonk::AUX_OFFSET,
```

Comment:
There is no need to setup aux_offset.  Just assign 0 to it.

Reply:
The problem probably is raised here, cause the CircomCuit we made is for Plonk, not for Groth.  Meanwhile, you have to fill in the witness here by `calculate_witness`.